### ISSUE TITLE
High memory usage on macos.

### ISSUE THREAD

**Issue by Lincest** [ahrm/sioyek#762]
## problem

I just opened a 15-page pdf and kept it for an hour, and the memory usage went from 200MB to 1.2GB.

![](https://youpai.roccoshi.top/img/202307100252548.png)

![](https://youpai.roccoshi.top/img/202307100252549.png)

## System information and installation method

- OS: `macOS 13.4 22F66 arm64`
- Installation: `brew install --cask sioyek`

**Comment by ahrm** [comment]
Can you share the PDF here?

**Comment by Lincest** [comment]
> Can you share the PDF here?

I don't think this error is related to a specific pdf, because I tried different pdfs today, all between 10-50 pages, but the memory usage is up to 1.9G, and it grows with time.

for example, file with 15 pages (1.3MB): 

[ElasticFlow.pdf](https://github.com/ahrm/sioyek/files/12005097/ElasticFlow.pdf)

**Comment by gogamid** [comment]
same here:
<img width="667" alt="image" src="https://github.com/ahrm/sioyek/assets/36050790/e9887b69-c30b-4179-b6d9-4525c7ae4947">

**Comment by ahrm** [comment]
This amount is normal, does it increase with time to unreasonable values? (e.g. more than 4gb)

**Comment by dbrll** [comment]
I wanted to report the same issue, Sioyek seems to use vast amounts of memory to open a PDF, sometimes more than 1GB. Apple's Preview uses a third of it. However its RSS is fairly small in comparison, about 56MB here.
Here are the two programs with the same short PDF opened:

<img width="638" alt="Screenshot 2023-07-15 at 09 43 36" src="https://github.com/ahrm/sioyek/assets/10487787/f6ce9ce8-fff7-4481-be0b-d84fda276206">

I suspect this may be caused by the lack of a native ARM64 build, the binary translation must add some overhead and resource usage. This may also explain why it's always using a couple percent CPU (and drain the battery) while Preview sits at 0%.

**Comment by ahrm** [comment]
Sioyek can be built natively for Arm64 already with no change to source code. The problem is, since I don't have a mac machine, I rely on github actions to build the release binaries and they don't have arm mac machines yet.

Anyway, I doubt the problem is even the architecture. Because of very high resolution displays on mac machines the bitmap of a single page can take upwards of 100MB, and we cache a few pages, which can consume ~1GB of memory, which I consider reasonable. I could make the cache size configurable in future versions so we use less memory for cache.

**Comment by zb-zombie** [comment]
I also met a high memory usage on windows. And maybe have scrolled too many pdfs, I got an "out of memory" error and then I rebooted machine. After that, I cannot make sioyek work any more, which means it just display a pure white board each time I open sioyek, and I cannot open the list of commands by pressing `:`, and I cannot shift open another pdf or open any new pdf (only pure white board is displayed). How can I make sioyek work again? @ahrm

**Comment by zb-zombie** [comment]
> How can I make sioyek work again?

After I reinstalled the sioyek and found it still didn't work, I realized there is something wrong with my computer, so I rebooted again and the problems gone.

**Comment by ideologysec** [comment]
Good news, arm64 macOS GitHub Runners are at least in public beta:

https://github.blog/2023-10-02-introducing-the-new-apple-silicon-powered-m1-macos-larger-runner-for-github-actions/

**Comment by miss4chewN8** [comment]
Want to bring up the same issue. Has the cache sizing been fixed?

<img width="957" alt="image" src="https://github.com/ahrm/sioyek/assets/76148233/4ab8e518-444b-402c-b0ad-a050163a59f5">

<img width="1730" alt="image" src="https://github.com/ahrm/sioyek/assets/76148233/9709425c-4c2f-4e31-bf34-b4b05c4b4401">

**Comment by kepler471** [comment]
Hey, I'm havng a similar issue. The memory grows directly with me scrolling through a pdf on the app.

Specs:
Macbook Pro M2 Pro
macOS Sonoma 14.3
sioyek version 2.0.0
```
~ » brew info sioyek
==> sioyek: 2.0.0
https://sioyek.info/
Installed
/opt/homebrew/Caskroom/sioyek/2.0.0 (221B)
From: https://github.com/Homebrew/homebrew-cask/blob/HEAD/Casks/s/sioyek.rb
==> Name
Sioyek
==> Description
PDF viewer designed for reading research papers and technical books
==> Artifacts
sioyek.app (App)
/opt/homebrew/Caskroom/sioyek/2.0.0/sioyek.wrapper.sh -> sioyek (Binary)
==> Analytics
install: 150 (30 days), 495 (90 days), 2,129 (365 days)
```

Video shows me scrolling through a 21 page pdf, with no notes, links or bookmarks.


https://github.com/user-attachments/assets/cfef75a0-6e55-4e53-9de8-e75f8e09e3b8

**Comment by ahrm** [comment]
Does the issue happen on the development branch?

**Comment by kepler471** [comment]
> Does the issue happen on the development branch?

I'll try it out and let you know :)

**Comment by kepler471** [comment]
> Does the issue happen on the development branch?

Sorry I've not been able to get the development branch to build on mac.
I think I'm getting issues with zsh, is there an alternative for the `setopt` line?
Also, do I need to use the alternate qt6 install? why cant I use the brew install?

**Comment by ahrm** [comment]
> is there a bash alternative for the setopt line?

Hmmm what `setopt` line?

> Also, do I need to use the alternate qt6 install? 

I am not sure, it might work with the brew install, I don't have a mac so I can't test that. You might need to install some additional qt library packages (at least that's the case in linux)

**Comment by KasselMarc** [comment]
Hey, I'm also experiencing similar high ram usage (I'm on the development branch). Is it already possible to configure the caching behaviour?

**Comment by ahrm** [comment]
You can change the `num_cached_pages` config (the default is 5), alternatively you can enable `sliced_rendering` which renders the pages in slices instead of rendering an entire page at once. This can reduce the memory load when only a portion of a page is visible.

**Comment by KasselMarc** [comment]
> You can change the `num_cached_pages` config (the default is 5), alternatively you can enable `sliced_rendering` which renders the pages in slices instead of rendering an entire page at once. This can reduce the memory load when only a portion of a page is visible.

Ok, thanks!

**Comment by zhizhongpu** [comment]
I experience a similar issue (alpha3 development branch). 

![Image](https://github.com/user-attachments/assets/5992c914-bae8-4165-8fc1-02e34f17c455)

> You can change the `num_cached_pages` config (the default is 5), alternatively you can enable `sliced_rendering` which renders the pages in slices instead of rendering an entire page at once. This can reduce the memory load when only a portion of a page is visible.

I understand the solution. Just reporting.

**Comment by giacomoborin** [comment]
I also have the same problem. Few hours ago I had to reboot the macbook since sioyek was using 90GBs of memory :(

**Comment by ahrm** [comment]
@giacomoborin  Which sioyek version are you using?

**Comment by giacomoborin** [comment]
v 2.0.0, installed with homebrew

**Comment by ahrm** [comment]
Does it happen with the preview release here: https://github.com/ahrm/sioyek/releases/tag/sioyek3-alpha0? (if you get an error message about file being damaged, see https://github.com/ahrm/sioyek/discussions/1156#discussioncomment-10822738)
